### PR TITLE
cpu/sam0_common/periph/uart: implement non-blocking write

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This replaces the polling based UART TX implementation on the sam0 platform with an interrupt based one.
Thread-safe Ringbuffer is used for the TX buffer with a compile-time configurable buffer size. (May be tweaked.)

~~**This violates the specification of `uart_write()`** as this makes [`uart_write()`](https://riot-os.org/api/group__drivers__periph__uart.html#gae2360fb0e880c387cfc02dd0cbd1c113) non-blocking.~~

The non-blocking mode is now only used if the module `periph_uart_nonblocking` is used.

I wanted to have a non-blocking debug output when debugging a race condition, so I implemented this. Maybe a non-blocking uart_write could be generally useful though.

### Testing procedure

add `USEMODULE += periph_uart_nonblocking` to your application.

Everything should behave as before, although the console output feels more snappy.
Only the shell / debug output have been tested.

### Issues/PRs references
none

### CODE REVIEW

**Comment by jcarrano** [comment]
I experimented with this a while ago, but never PR'd it. Here is my branch:

https://github.com/jcarrano/RIOT/commit/6e7e64a622975fc40a9a76390ed2568e92588b22

I defined a new function uart_write_async which blocks only if there is data waiting to be sent, otherwise it queues the current data and returns. I used a zero-copy approach by just storing a pointer to the buffer and then I redefined uart_write in terms of uart_write_async.

**Comment by benpicco** [comment]
I don't think zero-copy is the right approach for a general `uart_write` function though, e.g. you can't do zero-copy when using `printf`.

**Comment by dylad** [comment]
Isn't it better to create a dedicated pseudomodule for this ? the ROM cost looks heavy otherwise.
I can have a deeper look and run some test after the summit.

**Comment by jcarrano** [comment]
>  the ROM cost looks heavy otherwise.

Is there any application that stops fitting because of this? The cost of making it optional (at leat in the way it is done here and in most of riot) is a considerable increase in complexity and a decrease in readability, with all the ifdefs.

Add to that the fact that now you have two code paths in the uart driver that you have to test.

**Comment by keestux** [comment]
Is this a normal (allowed?) way of defining a regular macro?
```
USEMODULE += periph_uart_nonblocking
```
In most cases when I see `#ifdef MODULE_blabla` it means "if module blabla is present". But in this case, `periph_uart_nonblocking` is not a module. It is a macro to select a certain behaviour of the uart module.

Shouldn't we just have a macro (defined through CFLAGS) with a name such as `PERIPH_UART_NONBLOCKING`?

**Comment by keestux** [comment]
Notice that there is something called `PSEUDOMODULES`. I think makes sense to add `periph_uart_nonblocking` in `makefiles/pseudomodules.inc.mk`. No?

**Comment by benpicco** [comment]
There is already
```Makefile
PSEUDOMODULES += periph_%
```
in [`pseudomodules.inc.mk`](https://github.com/RIOT-OS/RIOT/blob/master/makefiles/pseudomodules.inc.mk#L155) - this also matches `periph_uart_nonblocking`.

**Comment by keestux** [comment]
@benpicco Yes, you're right. So it's not a problem.

File: Makefile.dep [hunk 1]
```diff
@@ -994,6 +994,10 @@ ifneq (,$(filter periph_%, $(USEMODULE)))
   USEMODULE += periph_common
 endif
 
+ifneq (,$(filter periph_uart_nonblocking, $(USEMODULE)))
+  USEMODULE += periph_uart
+endif
+
```
Comment (dylad) [inline Makefile.dep]: Why is this needed ?

**Comment by benpicco** [comment]
> any reason to keep SAM D&L 21 away from these changes ?

I also tested this on `samr21-xpro`, but only saml1x and samd5x have a separate TXE IRQ.

File: Makefile.dep (hunk 1 above)
Comment (benpicco) [reply Makefile.dep]: So everything that depends on `periph_uart` is also fulfilled with `periph_uart_nonblocking`

Comment (dylad) [reply Makefile.dep]: But should it belongs to Makefile.dep ? Why not cpu/sam0_common/Makefile.dep ?

Comment (benpicco) [reply Makefile.dep]: So other platforms can implement it too.

**Comment by benpicco** [comment]
> Remove all `cpu/samxx/Makefile.dep` introduced by this PR: they are empty and thus useless.

They are all but useless. This implementation depends on `tsrb`, this dependency is resolved in `sam0_common/Makefile.dep` so the sam0-siblings have to include it.

I've addressed all the other comments.
With the test application I get

```
== printed in 2100735/2100000 µs ==
```

with `periph_uart_nonblocking` and
```
== printed in 2152407/2100000 µs ==
```
with `periph_uart`.

**Comment by aabadie** [comment]
> They are all but useless.

You are right. I don't know why I thought they were initially empty.

Thanks for updating following the other suggestions. Just one is still missing: add a Python script for automatic testing.

**Comment by benpicco** [comment]
I added a small Python script.
When using `periph_uart` instead of `periph_uart_nonblocking` the test does fail as expected.

**Comment by aabadie** [comment]
@dylad, can you have a look at this one when you have time ? I'm fine with the test application and automatic test script. The build system integration is also ok for me.
Just the important part, e.g the non blocking uart code, needs a review now :)

File: cpu/sam0_common/periph/uart.c [hunk 2]
```diff
@@ -99,18 +113,26 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
     if ((rx_cb) && (uart_config[uart].rx_pin != GPIO_UNDEF)) {
         uart_ctx[uart].rx_cb = rx_cb;
         uart_ctx[uart].arg = arg;
-#if defined (CPU_SAML1X) || defined (CPU_SAMD5X)
+#ifdef UART_HAS_TX_ISR
+        /* enable RXNE ISR */
         NVIC_EnableIRQ(SERCOM0_2_IRQn + (sercom_id(dev(uart)) * 4));
 #else
+        /* enable UART ISR */
         NVIC_EnableIRQ(SERCOM0_IRQn + sercom_id(dev(uart)));
```
Comment (dylad) [inline cpu/sam0_common/periph/uart.c]: Got a hard time debugging this one.
In the current shape the provided test app doesn't work at all on SAML21.
No DRE interrupt is triggered. In fact, the NVIC_EnableIRQ is never call because their is no callback provided so we did not enter [here](https://github.com/RIOT-OS/RIOT/blob/7a16375438e960d46cecddae7e823c879d56a087/cpu/sam0_common/periph/uart.c#L113).
The callback is not provided because weren't using a shell.
So we need a second way to enable the NVIC in this specific case (TX only, no RX).

Comment (benpicco) [reply cpu/sam0_common/periph/uart.c]: Yea you are right.
I only tested the non-callback case on samd5x, for samd2x/saml2x the UART interrupt was only enabled when a callback was provided.
It is now also enabled when `MODULE_PERIPH_UART_NONBLOCKING` is defined.

File: cpu/sam0_common/periph/uart.c [hunk 3]
```diff
@@ -99,18 +113,31 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
         }
     }
+
+#ifdef MODULE_PERIPH_UART_NONBLOCKING
+#ifdef UART_HAS_TX_ISR
+    /* enable TXE ISR */
+    NVIC_EnableIRQ(SERCOM0_0_IRQn + (sercom_id(dev(uart)) * 4));
+#else
+    /* enable UART ISR */
+    NVIC_EnableIRQ(SERCOM0_IRQn + sercom_id(dev(uart)));
```
Comment (dylad) [inline cpu/sam0_common/periph/uart.c]: In some case, the NVIC_EnabeIRQ is called twice, this might be confusing.
I would rather go to something like :

```
    if ((rx_cb) && (uart_config[uart].rx_pin != GPIO_UNDEF)) {
        uart_ctx[uart].rx_cb = rx_cb;
        uart_ctx[uart].arg = arg;
#ifdef UART_HAS_TX_ISR
        /* enable RXNE ISR */
        NVIC_EnableIRQ(SERCOM0_2_IRQn + (sercom_id(dev(uart)) * 4));
#else
        /* enable UART ISR */
        NVIC_EnableIRQ(SERCOM0_IRQn + sercom_id(dev(uart)));
#endif /* UART_HAS_TX_ISR */
        dev(uart)->CTRLB.reg |= SERCOM_USART_CTRLB_RXEN;
        dev(uart)->INTENSET.reg |= SERCOM_USART_INTENSET_RXC;
        /* set wakeup receive from sleep if enabled */
        if (uart_config[uart].flags & UART_FLAG_WAKEUP) {
            dev(uart)->CTRLB.reg |= SERCOM_USART_CTRLB_SFDE;
        }
    }
#if defined(MODULE_PERIPH_UART_NONBLOCKING)
else {
    NVIC_EnableIRQ(SERCOM0_IRQn + sercom_id(dev(uart)));
}

#if defined(UART_HAS_TX_ISR) 
    /* enable TXE ISR */
    NVIC_EnableIRQ(SERCOM0_0_IRQn + (sercom_id(dev(uart)) * 4));
#endif
#endif /* MODULE_PERIPH_UART_NONBLOCKING */
```

**Comment by dylad** [comment]
This now works on SAML21 but I tested on SAML10 and SAML11 and here is result WITH uart_nonblocking:

```
2019-11-26 08:54:20,067 # m�ain(): This is RIOT! (Version: 2020.01-devel-426-ge36b5-sam0-buffe
red_uart)
2019-11-26 08:54:20,068 # 
2019-11-26 08:54:20,166 # J�oin us now and share the software;
2019-11-26 08:54:20,268 # Y�ou'll be free, hackers, you'll be free.
2019-11-26 08:54:20,368 # J�oin us now and share the software;
2019-11-26 08:54:20,470 # Y�ou'll be free, hackers, you'll be free.
2019-11-26 08:54:20,568 # 
2019-11-26 08:54:20,671 # �H�oarders can get piles of money,
2019-11-26 08:54:20,773 # T�hat is true, hackers, that is true.
2019-11-26 08:54:20,874 # B�ut they cannot help their neighbors;
2019-11-26 08:54:20,976 # T�hat's not good, hackers, that's not good.
2019-11-26 08:54:21,074 # 
2019-11-26 08:54:21,177 # �W�hen we have enough free software
2019-11-26 08:54:21,279 # A�t our call, hackers, at our call,
2019-11-26 08:54:21,380 # W�e'll kick out those dirty licenses
2019-11-26 08:54:21,481 # E�ver more, hackers, ever more.
2019-11-26 08:54:21,580 # 
2019-11-26 08:54:21,684 # �J�oin us now and share the software;
2019-11-26 08:54:21,785 # Y�ou'll be free, hackers, you'll be free.
2019-11-26 08:54:21,886 # J�oin us now and share the software;
2019-11-26 08:54:21,988 # Y�ou'll be free, hackers, you'll be free.
2019-11-26 08:54:22,085 # 
2019-11-26 08:54:22,190 # �=�= printed in 2105365/2100000 µs ==
2019-11-26 08:54:22,191 # [SUCCESS]
```

The output is fine without the uart_nonblocking module. I'll try to investigate further.

**Comment by dylad** [comment]
One really quick&dirty workaround is the following:

```
#ifdef MODULE_PERIPH_UART_NONBLOCKING
static inline void irq_handler_tx(unsigned uartnum)
{
    if(!tsrb_avail(&uart_tx_rb[uartnum])) {
        return;
    }
    dev(uartnum)->DATA.reg = tsrb_get_one(&uart_tx_rb[uartnum]);
    /* disable the interrupt if there are no more bytes to send */
    if (tsrb_empty(&uart_tx_rb[uartnum])) {
        dev(uartnum)->INTENCLR.reg = SERCOM_USART_INTENSET_DRE;
    }
}
#endif
```
But I'd like to find a better patch...

**Comment by dylad** [comment]
static tests failed, you can amend directly

**Comment by dylad** [comment]
Here we go ! Thanks for the PR @benpicco !

**Comment by benpicco** [comment]
Thank you for the review!
